### PR TITLE
Fix formatter soundness check

### PR DESCRIPTION
**Pull request by shaobo-he-aws**
## Description of changes

There were two issues with the original function.

1. It iterates over templates instead of policies.
2. It zips over iterations coming from hashmaps, whose orders are non-deterministic.

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A bug fix or other functionality change requiring a patch to `cedar-policy`.

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

### CODE REVIEW

**Review by john-h-kastner-aws — APPROVED** [review]
Can we add a test case?

**Comment by shaobo-he-aws** [comment]
> Can we add a test case?

Good idea. Working on it.

File: cedar-policy-formatter/src/pprint/fmt.rs [hunk 1]
```diff
@@ -135,6 +143,49 @@ mod tests {
...
+        let p1 = r#"permit (principal, action, resource)
+        when { "
+        
+        a
+        " };"#;
+        let p2 = r#"permit (principal, action, resource)
+        when { "
+        a
+        " };"#;
+        assert!(soundness_check(p2, &parse_policyset(p1).unwrap()).is_err());
```
Comment (khieta) [inline cedar-policy-formatter/src/pprint/fmt.rs:157]: Probably a dumb question, but why does the change in this PR help identify the issue in #862? Looking at the diff, it seems like the only change is to add a comparison of policy ids, which doesn't seem related.

Or is this test case unrelated to the changed code?

File: cedar-policy-formatter/src/pprint/fmt.rs [hunk 2]
```diff
@@ -69,7 +77,7 @@ fn soundness_check(ps: &str, ast: &PolicySet) -> Result<()> {
                 .eq_shape(p.non_scope_constraints()))
         {
             return Err(miette!(
-                "policies differ in meaning or annotations:\noriginal: {p}\nformatted: {f_p}"
+                "policies differ in policy ids or meaning or annotations:\noriginal: {p}\nformatted: {f_p}"
```
Comment (khieta) [inline cedar-policy-formatter/src/pprint/fmt.rs:80]: Nit: probably easier for debugging if you separate these cases.

File: cedar-policy-formatter/src/pprint/fmt.rs (hunk 1 above)
Comment (john-h-kastner-aws) [reply cedar-policy-formatter/src/pprint/fmt.rs:157]: `templates()` to `policies()` I think

Comment (khieta) [reply cedar-policy-formatter/src/pprint/fmt.rs:157]: Aha, thanks. @shaobo-he-aws can you add a description to the PR so it's easy not to miss this.

Comment (shaobo-he-aws) [reply cedar-policy-formatter/src/pprint/fmt.rs:157]: Done.

File: cedar-policy-formatter/src/pprint/fmt.rs (hunk 2 above)
Comment (shaobo-he-aws) [reply cedar-policy-formatter/src/pprint/fmt.rs:80]: I think I'll leave it to future work
